### PR TITLE
Add CUDA matrix-free operator

### PR DESCRIPTION
**Pull request by Rombur**
This is almost identical to the matrix-free operator on the host.

### CODE REVIEW

File: source/cuda/cuda_matrix_free_operator.cu [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  // FIXME Return garbage since throwing not implemented will make important
+  // tests to fail
+  return 0;
+}
+
+template <int dim, typename VectorType>
+size_t CudaMatrixFreeOperator<dim, VectorType>::operator_complexity() const
+{
+  // FIXME Return garbage since throwing not implemented will make important
+  // tests to fail
```
Comment (masterleinad) [inline source/cuda/cuda_matrix_free_operator.cu]: Which test would fail?

Comment (Rombur) [reply source/cuda/cuda_matrix_free_operator.cu]: `test_hierarchy_device` I guess we can also remove the line in that output the grid and the operator complexity like we did on the host where we hit the same problem.

File: source/cuda/cuda_matrix_free_operator.cu [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  ASSERT_CUSPARSE(cusparse_error_code);
+  cusparse_error_code =
+      cusparseSetMatType(c_dev->descr, CUSPARSE_MATRIX_TYPE_GENERAL);
+  ASSERT_CUSPARSE(cusparse_error_code);
+  cusparse_error_code =
+      cusparseSetMatIndexBase(c_dev->descr, CUSPARSE_INDEX_BASE_ZERO);
+  ASSERT_CUSPARSE(cusparse_error_code);
+
+  std::shared_ptr<Operator<VectorType>> op(
+      new CudaMatrixOperator<VectorType>(c_dev));
```
Comment (masterleinad) [inline source/cuda/cuda_matrix_free_operator.cu]: You should be able to just do
```
return std::make_shared<CudaMatrixOperator<VectorType>>(c_dev);
``` 
here.

File: source/cuda/cuda_matrix_free_operator.cu [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+template <int dim, typename VectorType>
+std::shared_ptr<VectorType>
+CudaMatrixFreeOperator<dim, VectorType>::build_domain_vector() const
+{
+  // We know the operator is squared
+  auto domain_vector = this->build_range_vector();
+
+  return domain_vector;
```
Comment (masterleinad) [inline source/cuda/cuda_matrix_free_operator.cu]: You can also just compress this to
```
return this->build_range_vector();
```

File: include/mfmg/cuda/cuda_matrix_free_operator.cuh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  // Needed because the extract_row works only on the host
+  void apply(dealii::LinearAlgebra::distributed::Vector<
+                 value_type, dealii::MemorySpace::Host> const &x,
+             dealii::LinearAlgebra::distributed::Vector<
+                 value_type, dealii::MemorySpace::Host> &y,
+             OperatorMode mode = OperatorMode::NO_TRANS) const;
+
+  virtual void apply(vector_type const &x, vector_type &y,
+                     OperatorMode mode = OperatorMode::NO_TRANS) const override;
```
Comment (dalg24) [inline include/mfmg/cuda/cuda_matrix_free_operator.cuh]: I am getting confused but the coexistence of `vmult()` and `apply()` overload that take `vector_type` and the `apply()` overload that takes vectors on the host.  Was the latter introduced so you can reuse `matrix_transpose_matrix_multiply()`?

Comment (Rombur) [reply include/mfmg/cuda/cuda_matrix_free_operator.cuh]: Yes exactly. I'll add a comment.

File: source/cuda/cuda_matrix_free_operator.cu (hunk 2 above)
Comment (Rombur) [reply source/cuda/cuda_matrix_free_operator.cu]: meh that's just a stylistic difference

**Comment by Rombur** [comment]
I have added a comment about the different `apply()`. I haven't done anything for `operator_complexity()` and `grid_complexity()` because they will disappear in the future.

**Comment by Rombur** [comment]
_ping_

**Comment by Rombur** [comment]
_ping_

**Pull request merged**